PyTorchLightning/pytorch-lightning-bolts
# ci: Cache datasets

## What does this PR do?
Fixes #279. All tests now use `./datasets/` to store downloaded datasets and to restore from if already downloaded.

### Description of the changes
- GitHub Actions
  - Add `Cache datasets` section
- pytest
  - Remove unused `tmpdir` fixture
  - Create `tests/conftest.py` to define `data_dir` and `DATA_DIR` which can be referenced from each test.
    - `data_dir` fixture can be used without any import in each test.
    - `DATA_DIR` const variable is defined to refer to the dataset directory in `pytest.mark.parametrize()`. pytest does not support using fixtures in `pytest.mark.parametrize()`. See https://github.com/pytest-dev/pytest/issues/349.

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃
